### ISSUE TITLE
Combining both depth and pose net parameters in a single optimizer instead of a separate ones.

### ISSUE THREAD

**Issue by resindraburiza** [nianticlabs/monodepth2#304]
Hello,

this is more of a general discussion/question rather than an issue. In monodepth2 implementation, there are two separate networks: depth prediction and pose prediction network. To optimize the network, monodepth2 uses a single optimizer instance with both network parameters as input. Are there any considerations to using a single optimizer instead of separate optimizers for each network? Will there be any effects? When do we should use separate optimizer for each network instead of a single optimizer?

I am sorry if it is a trivial question. It would be very helpful if someone could guide me regarding this matter. Thank you!

**Comment by mdfirman** [comment]
Thanks for the interesting question. In my view it would only make sense to have a separate optimiser for each network if each network had a separate goal to achieve.

In our case, the two networks both share the same goal, i.e. to find the lowest image reconstruction loss. Because the two networks share the same common goal, they should use the same optimiser. We could try to do it with separate optimisers, but I doubt it would work as well. It would be an interesting experiment to try though.

**Comment by resindraburiza** [comment]
Thank you very much for your reply. 
I would like to discuss it further if I may. Based on what you have explained, the goal can be defined as a loss function. So, as long as there is one final loss function,  all the network parameters can be combined into a single optimizer. Is this a correct understanding?

Let say, I have depth ground truth data and I want to add direct depth supervision to the network defined as L_depth. Then, the final loss including the original monodepth2 loss can be written as

L_total = w1 * L_depth + w2 * L_photometric + w3 * L_smoothness
where L_depth is L1 difference between predicted depth and depth ground truth.

as long as we use the correct weight, can we still use a single optimizer for both depth and pose network? 
Thank you so much.

**Comment by mdfirman** [comment]
My answers to your questions are: yes, and yes!

In general I think it is better to use one optimiser and train everything together. The optimiser's job is to make the loss as low as possible, and it is able to do so by modifying all the trainable parameters it has access to. In theory one single optimiser may be able to find a better minimum than using separate optimisers for each network, which each only get to see a subset of the trainable parameters, i.e. a blinkered view.

**Comment by resindraburiza** [comment]
Thank you so much for your time to answer my simple question. I learn something from this discussion.
